File: src/qjson.py
# ============================================================
# qjson.py — QJSON: JSON + comments + BigInt + BigDecimal + BigFloat
#
# Superset of JSON using QuickJS bignum syntax:
#   123N          → BigInt      (Python: BigInt subclass of int)
#   123.456M      → BigDecimal  (Python: decimal.Decimal)
#   3.14L         → BigFloat    (Python: BigFloat — preserves full precision)
#
# Uppercase preferred, lowercase accepted.  Consistent and visible.
#   // line       → comment
#   /* block */   → comment
#
# Valid JSON is valid QJSON.  No collisions.
#
# Usage:
#   from qjson import parse, stringify, BigInt, BigFloat
#   obj = parse('{"n": 42n, "d": 3.14m, "f": 3.14l}')
#   text = stringify(obj)
# ============================================================

from decimal import Decimal


# ── JS64 blob encoding ──────────────────────────────────────

_JS64_ALPHA = "$0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz"
_JS64_REV = None


def _js64_init_rev():
    global _JS64_REV
    _JS64_REV = {}
    for i, c in enumerate(_JS64_ALPHA):
        _JS64_REV[c] = i


def js64_decode(s):
    """Decode JS64 string (without leading '$') to bytes."""
    if _JS64_REV is None:
        _js64_init_rev()
    js64 = "$" + s  # restore the implicit leading '$'
    js64_len = len(js64) - 1
    blob_len = (js64_len * 3) >> 2
    blob = bytearray(blob_len)
    code = 0
    bits = 0
    byte_idx = 0
    for i in range(js64_len):
        v = _JS64_REV.get(js64[i + 1])
        if v is None:
            raise ValueError("Invalid JS64 character: %r" % js64[i + 1])
        code = code | (v << bits)
        bits += 6
        if bits >= 8:
            if byte_idx < blob_len:
                blob[byte_idx] = code & 0xFF
            code = code >> 8
            bits -= 8
            byte_idx += 1
    return bytes(blob)


def js64_encode(data):
    """Encode bytes to JS64 string (without leading '$', for 0j prefix)."""
    if isinstance(data, (bytes, bytearray)):
        data = list(data)
    js64_len = ((len(data) * 4 + 2) // 3)
    parts = []
    code = 0
    bits = 6  # start with 6 zero bits (the implicit '$')
    byte_idx = 0
    for i in range(js64_len + 1):
        ch = _JS64_ALPHA[code & 0x3F]
        if i > 0:  # skip the leading '$'
            parts.append(ch)
        code = code >> 6
        bits -= 6
        if bits < 6 and byte_idx < len(data):
            code = code | (data[byte_idx] << bits)
            bits += 8
            byte_idx += 1
    return "".join(parts)


class Blob:
    """Binary data that round-trips through QJSON with 0j prefix (JS64)."""
    __slots__ = ("data",)

    def __init__(self, data):
        if isinstance(data, (list, tuple)):
            data = bytes(data)
        self.data = bytes(data)

    def __eq__(self, other):
        if isinstance(other, Blob):
            return self.data == other.data
        return NotImplemented

    def __hash__(self):
        return hash(("Blob", self.data))

    def __repr__(self):
        return "Blob(%r)" % self.data


class BigInt(int):
    """Integer that round-trips through QJSON with 'n' suffix."""
    pass


class BigFloat:
    """High-precision base-2 float.  Round-trips with 'l' suffix.

    Stores the full-precision string so no bits are lost.
    float(bf) gives the nearest 64-bit IEEE value.
    """
    __slots__ = ("_raw",)

    def __init__(self, value):
        self._raw = str(value)

    def __float__(self):
        return float(self._raw)

    def __repr__(self):
        return "BigFloat('%s')" % self._raw

    def __str__(self):
        return self._raw

    def __eq__(self, other):
        if isinstance(other, BigFloat):
            return self._raw == other._raw
        return NotImplemented

    def __hash__(self):
        return hash(("BigFloat", self._raw))


# ── Parser ───────────────────────────────────────────────────

def parse(text):
    """Parse QJSON text to Python objects."""
    p = _Parser(text)
    val = p.value()
    p.ws()
    if p.pos < p.end:
        raise ValueError("Trailing content at %d" % p.pos)
    return val


class _Parser:
    __slots__ = ("text", "pos", "end")

    def __init__(self, text):
        self.text = text
        self.pos = 0
        self.end = len(text)

    def ch(self):
        return self.text[self.pos] if self.pos < self.end else ""

    def ws(self):
        while self.pos < self.end:
            c = self.text[self.pos]
            if c in " \t\n\r":
                self.pos += 1
            elif c == "/" and self.pos + 1 < self.end:
                c2 = self.text[self.pos + 1]
                if c2 == "/":
                    self.pos += 2
                    while self.pos < self.end and self.text[self.pos] != "\n":
                        self.pos += 1
                elif c2 == "*":
                    self.pos += 2
                    depth = 1
                    while self.pos + 1 < self.end and depth > 0:
                        if self.text[self.pos] == "/" and self.text[self.pos + 1] == "*":
                            depth += 1
                            self.pos += 2
                        elif self.text[self.pos] == "*" and self.text[self.pos + 1] == "/":
                            depth -= 1
                            self.pos += 2
                        else:
                            self.pos += 1
                    if depth > 0:
                        raise ValueError("Unterminated block comment")
                else:
                    break
            else:
                break

    def expect(self, c):
        if self.pos >= self.end or self.text[self.pos] != c:
            raise ValueError("Expected '%s' at %d" % (c, self.pos))
        self.pos += 1

    def ident(self):
        """Parse an unquoted key (JS identifier)."""
        start = self.pos
        c = self.ch()
        if not (c.isalpha() or c == "_" or c == "$"):
            raise ValueError("Expected identifier at %d" % self.pos)
        self.pos += 1
        while self.pos < self.end:
            c = self.text[self.pos]
            if c.isalnum() or c == "_" or c == "$":
                self.pos += 1
            else:
                break
        return self.text[start:self.pos]

    def key(self):
        """Parse a key: quoted string or bare identifier."""
        if self.ch() == '"':
            return self.string()
        return self.ident()

    def value(self):
        self.ws()
        c = self.ch()
        if c == '"':  return self.string()
        if c == "{":  return self.obj()
        if c == "[":  return self.arr()
        if c == "t":  return self.literal("true", True)
        if c == "f":  return self.literal("false", False)
        if c == "n" and self.text[self.pos:self.pos + 4] == "null":
            return self.literal("null", None)
        if c == "0" and self.pos + 1 < self.end and self.text[self.pos + 1] in "jJ":
            return self.blob()
        if c == "-" or c.isdigit():
            return self.number()
        raise ValueError("Unexpected '%s' at %d" % (c, self.pos))

    def literal(self, word, val):
        if self.text[self.pos:self.pos + len(word)] != word:
            raise ValueError("Expected '%s' at %d" % (word, self.pos))
        self.pos += len(word)
        return val

    def string(self):
        self.expect('"')
        parts = []
        while self.pos < self.end:
            c = self.text[self.pos]
            if c == '"':
                self.pos += 1
                return "".join(parts)
            if c == "\\":
                self.pos += 1
                e = self.text[self.pos]
                if   e == '"':  parts.append('"')
                elif e == "\\": parts.append("\\")
                elif e == "/":  parts.append("/")
                elif e == "b":  parts.append("\b")
                elif e == "f":  parts.append("\f")
                elif e == "n":  parts.append("\n")
                elif e == "r":  parts.append("\r")
                elif e == "t":  parts.append("\t")
                elif e == "u":
                    h = self.text[self.pos + 1:self.pos + 5]
                    parts.append(chr(int(h, 16)))
                    self.pos += 4
                self.pos += 1
            else:
                parts.append(c)
                self.pos += 1
        raise ValueError("Unterminated string")

    def number(self):
        start = self.pos
        if self.ch() == "-":
            self.pos += 1
        while self.pos < self.end and self.text[self.pos].isdigit():
            self.pos += 1
        is_float = False
        if self.pos < self.end and self.text[self.pos] == ".":
            is_float = True
            self.pos += 1
            while self.pos < self.end and self.text[self.pos].isdigit():
                self.pos += 1
        if self.pos < self.end and self.text[self.pos] in "eE":
            is_float = True
            self.pos += 1
            if self.pos < self.end and self.text[self.pos] in "+-":
                self.pos += 1
            while self.pos < self.end and self.text[self.pos].isdigit():
                self.pos += 1
        raw = self.text[start:self.pos]
        # BigInt suffix (N preferred, n accepted)
        if self.pos < self.end and self.text[self.pos] in "nN":
            self.pos += 1
            return BigInt(raw)
        # BigDecimal suffix (M preferred, m accepted)
        if self.pos < self.end and self.text[self.pos] in "mM":
            self.pos += 1
            return Decimal(raw)
        # BigFloat suffix (L preferred, l accepted)
        if self.pos < self.end and self.text[self.pos] in "lL":
            self.pos += 1
            return BigFloat(raw)
        # Regular number
        if is_float:
            return float(raw)
        return int(raw)

    def blob(self):
        self.pos += 2  # skip 0j / 0J
        start = self.pos
        js64_chars = set(_JS64_ALPHA)
        while self.pos < self.end and self.text[self.pos] in js64_chars:
            self.pos += 1
        raw = self.text[start:self.pos]
        return Blob(js64_decode(raw))

    def obj(self):
        self.expect("{")
        d = {}
        self.ws()
        if self.ch() == "}":
            self.pos += 1
            return d
        while True:
            self.ws()
            k = self.key()
            self.ws()
            self.expect(":")
            d[k] = self.value()
            self.ws()
            if self.ch() == "}":
                self.pos += 1
                return d
            self.expect(",")
            self.ws()
            if self.ch() == "}":  # trailing comma
                self.pos += 1
                return d

    def arr(self):
        self.expect("[")
        a = []
        self.ws()
        if self.ch() == "]":
            self.pos += 1
            return a
        while True:
            a.append(self.value())
            self.ws()
            if self.ch() == "]":
                self.pos += 1
                return a
            self.expect(",")
            self.ws()
            if self.ch() == "]":  # trailing comma
                self.pos += 1
                return a


# ── Serializer ───────────────────────────────────────────────

def stringify(obj, indent=None):
    """Serialize to QJSON.  BigInt → 'n', Decimal → 'm', BigFloat → 'l'."""
    return _fmt(obj, indent, 0)


def _fmt(obj, ind, depth):
    if obj is None:
        return "null"
    if obj is True:
        return "true"
    if obj is False:
        return "false"
    if isinstance(obj, Blob):
        return "0j" + js64_encode(obj.data)
    if isinstance(obj, BigFloat):
        return obj._raw + "L"
    if isinstance(obj, BigInt):
        return int.__repr__(obj) + "N"
    if isinstance(obj, Decimal):
        return str(obj) + "M"
    if isinstance(obj, float):
        if obj != obj or obj == float("inf") or obj == float("-inf"):
            return "null"
        return repr(obj)
    if isinstance(obj, int):
        return str(obj)
    if isinstance(obj, str):
        return _esc(obj)
    if isinstance(obj, (list, tuple)):
        if not obj:
            return "[]"
        if ind is None:
            return "[" + ",".join(_fmt(v, None, 0) for v in obj) + "]"
        nl = "\n" + " " * (ind * (depth + 1))
        end = "\n" + " " * (ind * depth)
        return "[" + ",".join(nl + _fmt(v, ind, depth + 1) for v in obj) + end + "]"
    if isinstance(obj, dict):
        if not obj:
            return "{}"
        if ind is None:
            return "{" + ",".join(
                _esc(k) + ":" + _fmt(v, None, 0) for k, v in obj.items()
            ) + "}"
        nl = "\n" + " " * (ind * (depth + 1))
        end = "\n" + " " * (ind * depth)
        return "{" + ",".join(
            nl + _esc(k) + ": " + _fmt(v, ind, depth + 1) for k, v in obj.items()
        ) + end + "}"
    return str(obj)


_ESC = {'"': '\\"', "\\": "\\\\", "\n": "\\n", "\r": "\\r",
        "\t": "\\t", "\b": "\\b", "\f": "\\f"}

def _esc(s):
    r = ['"']
    for c in s:
        if c in _ESC:
            r.append(_ESC[c])
        elif ord(c) < 0x20:
            r.append("\\u%04x" % ord(c))
        else:
            r.append(c)
    r.append('"')
    return "".join(r)

File: src/qjson.js
// ============================================================
// qjson.js — QJSON: JSON + comments + BigInt + BigDecimal + BigFloat
//
// Superset of JSON using QuickJS bignum syntax:
//   123N          → BigInt      (native if available, else string)
//   123.456M      → BigDecimal  (QuickJS native, else string)
//   3.14L         → BigFloat    (QuickJS native, else string)
//
// Uppercase preferred, lowercase accepted.  Consistent and visible.
//   // line       → comment
//   /* block */   → comment
//
// Valid JSON is valid QJSON.  No collisions.
//
// Portable: ES5 style (var, function, no arrows).
// BigInt/BigDecimal used only at runtime if the host supports them.
//
// Usage:
//   var q = qjson.parse('{"n": 42n, "d": 3.14m}');
//   var s = qjson.stringify(q);
// ============================================================

// ── JS64 blob encoding ──────────────────────────────────────

var _js64alpha = "$0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz";
var _js64rev = null;

function _js64_init_rev() {
  _js64rev = {};
  for (var i = 0; i < _js64alpha.length; i++) {
    _js64rev[_js64alpha.charAt(i)] = i;
  }
}

function js64_decode(str) {
  // str does NOT have leading '$' — prepend it for decoding
  if (_js64rev === null) _js64_init_rev();
  var js64 = "$" + str;
  var js64len = js64.length - 1;
  var blobLen = (js64len * 3) >> 2;
  var blob = [];
  var code = 0, bits = 0, byte = 0;
  for (var i = 0; i < js64len; i++) {
    var v = _js64rev[js64.charAt(i + 1)];
    if (v === undefined) throw new Error("Invalid JS64 character");
    code = code | (v << bits);
    bits += 6;
    if (bits >= 8) {
      blob[byte] = code & 0xFF;
      code = code >> 8;
      bits -= 8;
      byte++;
    }
  }
  return blob;
}

function js64_encode(blob) {
  // Returns string WITHOUT leading '$' (for 0j prefix)
  var js64len = ((blob.length * 4 + 2) / 3) | 0;
  var parts = [];
  var code = 0, bits = 6, byte = 0;
  for (var i = 0; i <= js64len; i++) {
    var ch = _js64alpha.charAt(code & 0x3F);
    if (i > 0) parts.push(ch); // skip the leading '$'
    code = code >> 6;
    bits -= 6;
    if (bits < 6 && byte < blob.length) {
      code = code | (blob[byte] << bits);
      bits += 8;
      byte++;
    }
  }
  return parts.join("");
}

// ── Parser ──────────────────────────────────────────────────

function qjson_parse(text) {
  var pos = 0;
  var len = text.length;

  function ch() { return pos < len ? text[pos] : ""; }

  function ws() {
    while (pos < len) {
      var c = text[pos];
      if (c === " " || c === "\t" || c === "\n" || c === "\r") { pos++; continue; }
      if (c === "/" && pos + 1 < len) {
        if (text[pos + 1] === "/") {
          pos += 2;
          while (pos < len && text[pos] !== "\n") pos++;
          continue;
        }
        if (text[pos + 1] === "*") {
          pos += 2;
          var depth = 1;
          while (pos + 1 < len && depth > 0) {
            if (text[pos] === "/" && text[pos + 1] === "*") { depth++; pos += 2; }
            else if (text[pos] === "*" && text[pos + 1] === "/") { depth--; pos += 2; }
            else pos++;
          }
          continue;
        }
      }
      break;
    }
  }

  function expect(c) {
    if (pos >= len || text[pos] !== c) throw new Error("Expected '" + c + "' at " + pos);
    pos++;
  }

  function ident() {
    var start = pos;
    var c = ch();
    if (!((c >= "a" && c <= "z") || (c >= "A" && c <= "Z") || c === "_" || c === "$"))
      throw new Error("Expected identifier at " + pos);
    pos++;
    while (pos < len) {
      c = text[pos];
      if ((c >= "a" && c <= "z") || (c >= "A" && c <= "Z") ||
          (c >= "0" && c <= "9") || c === "_" || c === "$") pos++;
      else break;
    }
    return text.substring(start, pos);
  }

  function key() {
    if (ch() === '"') return string();
    return ident();
  }

  function value() {
    ws();
    var c = ch();
    if (c === '"') return string();
    if (c === "{") return obj();
    if (c === "[") return arr();
    if (c === "t") return literal("true", true);
    if (c === "f") return literal("false", false);
    if (c === "n" && text.substr(pos, 4) === "null") return literal("null", null);
    if (c === "0" && pos + 1 < len && (text[pos + 1] === "j" || text[pos + 1] === "J")) return blob();
    if (c === "-" || (c >= "0" && c <= "9")) return number();
    throw new Error("Unexpected '" + c + "' at " + pos);
  }

  function literal(word, val) {
    if (text.substr(pos, word.length) !== word) throw new Error("Expected '" + word + "' at " + pos);
    pos += word.length;
    return val;
  }

  function string() {
    expect('"');
    var parts = [];
    while (pos < len) {
      var c = text[pos];
      if (c === '"') { pos++; return parts.join(""); }
      if (c === "\\") {
        pos++;
        var e = text[pos];
        if      (e === '"')  parts.push('"');
        else if (e === "\\") parts.push("\\");
        else if (e === "/")  parts.push("/");
        else if (e === "b")  parts.push("\b");
        else if (e === "f")  parts.push("\f");
        else if (e === "n")  parts.push("\n");
        else if (e === "r")  parts.push("\r");
        else if (e === "t")  parts.push("\t");
        else if (e === "u") {
          parts.push(String.fromCharCode(parseInt(text.substr(pos + 1, 4), 16)));
          pos += 4;
        }
        pos++;
      } else {
        parts.push(c);
        pos++;
      }
    }
    throw new Error("Unterminated string");
  }

  function number() {
    var start = pos;
    if (ch() === "-") pos++;
    while (pos < len && text[pos] >= "0" && text[pos] <= "9") pos++;
    var isFloat = false;
    if (pos < len && text[pos] === ".") {
      isFloat = true;
      pos++;
      while (pos < len && text[pos] >= "0" && text[pos] <= "9") pos++;
    }
    if (pos < len && (text[pos] === "e" || text[pos] === "E")) {
      isFloat = true;
      pos++;
      if (pos < len && (text[pos] === "+" || text[pos] === "-")) pos++;
      while (pos < len && text[pos] >= "0" && text[pos] <= "9") pos++;
    }
    var raw = text.substring(start, pos);
    // BigInt suffix (N preferred, n accepted)
    if (pos < len && (text[pos] === "N" || text[pos] === "n")) {
      pos++;
      if (typeof BigInt !== "undefined") return BigInt(raw);
      return raw + "N";
    }
    // BigDecimal suffix (M preferred, m accepted)
    if (pos < len && (text[pos] === "M" || text[pos] === "m")) {
      pos++;
      if (typeof BigDecimal !== "undefined") return BigDecimal(raw);
      return raw + "M";
    }
    // BigFloat suffix (L preferred, l accepted)
    if (pos < len && (text[pos] === "L" || text[pos] === "l")) {
      pos++;
      if (typeof BigFloat !== "undefined") return BigFloat(raw);
      return raw + "L";
    }
    // Regular number
    if (isFloat) return parseFloat(raw);
    var n = parseInt(raw, 10);
    return n;
  }

  function blob() {
    pos += 2; // skip 0j / 0J
    var start = pos;
    while (pos < len) {
      var c = text[pos];
      if ((c >= "0" && c <= "9") || (c >= "A" && c <= "Z") ||
          (c >= "a" && c <= "z") || c === "$" || c === "_") pos++;
      else break;
    }
    var raw = text.substring(start, pos);
    return { $qjson: "blob", data: js64_decode(raw) };
  }

  function obj() {
    expect("{");
    var d = {};
    ws();
    if (ch() === "}") { pos++; return d; }
    while (true) {
      ws();
      var k = key();
      ws(); expect(":");
      d[k] = value();
      ws();
      if (ch() === "}") { pos++; return d; }
      expect(",");
      ws();
      if (ch() === "}") { pos++; return d; }  // trailing comma
    }
  }

  function arr() {
    expect("[");
    var a = [];
    ws();
    if (ch() === "]") { pos++; return a; }
    while (true) {
      a.push(value());
      ws();
      if (ch() === "]") { pos++; return a; }
      expect(",");
      ws();
      if (ch() === "]") { pos++; return a; }  // trailing comma
    }
  }

  var result = value();
  ws();
  if (pos < len) throw new Error("Trailing content at " + pos);
  return result;
}

// ── Serializer ──────────────────────────────────────────────

function qjson_stringify(obj) {
  return _fmt(obj);
}

function _fmt(obj) {
  if (obj === null || obj === undefined) return "null";
  if (obj === true)  return "true";
  if (obj === false) return "false";
  // BigInt
  if (typeof obj === "bigint") return String(obj) + "N";
  // BigDecimal (QuickJS)
  if (typeof obj === "bigdecimal") return obj.toString() + "M";
  // BigFloat (QuickJS)
  if (typeof obj === "bigfloat") return obj.toString() + "L";
  // Number
  if (typeof obj === "number") {
    if (obj !== obj || obj === Infinity || obj === -Infinity) return "null";
    return String(obj);
  }
  // Blob
  if (typeof obj === "object" && obj !== null && obj.$qjson === "blob") {
    return "0j" + js64_encode(obj.data);
  }
  // String
  if (typeof obj === "string") return _esc(obj);
  // Array
  if (Array.isArray(obj)) {
    if (obj.length === 0) return "[]";
    var items = [];
    for (var i = 0; i < obj.length; i++) items.push(_fmt(obj[i]));
    return "[" + items.join(",") + "]";
  }
  // Object
  if (typeof obj === "object") {
    var keys = Object.keys(obj);
    if (keys.length === 0) return "{}";
    var pairs = [];
    for (var i = 0; i < keys.length; i++) {
      pairs.push(_esc(keys[i]) + ":" + _fmt(obj[keys[i]]));
    }
    return "{" + pairs.join(",") + "}";
  }
  return String(obj);
}

function _esc(s) {
  var r = '"';
  for (var i = 0; i < s.length; i++) {
    var c = s[i];
    if      (c === '"')  r += '\\"';
    else if (c === "\\") r += "\\\\";
    else if (c === "\n") r += "\\n";
    else if (c === "\r") r += "\\r";
    else if (c === "\t") r += "\\t";
    else if (c === "\b") r += "\\b";
    else if (c === "\f") r += "\\f";
    else if (c.charCodeAt(0) < 0x20) {
      var h = c.charCodeAt(0).toString(16);
      r += "\\u" + ("0000" + h).slice(-4);
    }
    else r += c;
  }
  return r + '"';
}

// ── Export (dual ESM/CJS) ───────────────────────────────────

if (typeof exports !== "undefined") {
  exports.qjson_parse = qjson_parse;
  exports.qjson_stringify = qjson_stringify;
  exports.js64_encode = js64_encode;
  exports.js64_decode = js64_decode;
}
export { qjson_parse, qjson_stringify, js64_encode, js64_decode };
